### PR TITLE
Add age-varying and time-varying data simulation including seroreversion 

### LINKED ISSUE #74: Simulate from age-varying FOI serocatalytic model
Add a feature and functionality for age varying FOI model: 
```math

\displaylines{ 
&lambda;(a) = &lambda;_{0} e^{-ra} \ \
dS/da = -&lambda;S + &mu;Z \ \
dZ/da = &lambda;S - &mu;Z \ \
dZ/da = &lambda;(a)(1-Z) - &mu;Z \ \

Z = exp(- \sum_{{a\prime}=0}^a [&mu; + &lambda;({a\prime})]) * \sum_{{a\prime\prime}=0}^a exp( \sum_{{a\prime}=0}^{a\prime\prime} [&mu; + &lambda;({a\prime})]) &lambda;({a\prime\prime})
 }

```
With a binomial likelihood:
```math
X_{i} &sim; B(N_{i}, Z_{i})

```

### PR DESCRIPTION
**Pull request by ntorresd**
This PR modifies `get_sim_probability` to allow for data simulation from age-varying FoI with seroreversion and includes seroreversion to the time-varying FoI option (@ben18785 please check if I'm computing this correctly). 

- The Booleans `is_time_varying` and `is_age_varying` can be used to specify whether age-varying or time-varying FoI is being used, with the time varying option being `TRUE` as default. 
- If both Booleans are set to `TRUE`, the user is warned about the lack of time and age-varying data simulation functionalities and the age varying option is selected. 
- If both Booleans are set to `FALSE`, the user is warned and the time-varying option is selected.
- The parameter `mu` corresponds to the seroreversion rate. If set to 0 (default), there's no seroreversion.

This PR closes #74 (I'll move the discussion about the age-varying models to a separate issue).

### CODE REVIEW

File: R/seroprevalence_data.R [hunk 1]
```diff
@@ -139,6 +139,32 @@ prepare_bin_data <- function(serodata) {
...
+    mu = 0
+) {
+  # TODO: Avoid for loop if possible
+  probability <- 0
+  # solves ODE exactly within pieces
+  for (i in 1:exposure_age) {
+    probability <-
+      (1 / (foi[i] + mu)) * exp(- (foi[i] + mu)) *
+      (foi[i] * (exp(foi[i] + mu)  - 1) + probability * (foi[i] + mu)
+      )
```
Comment (ben18785) [inline R/seroprevalence_data.R:163]: Sorry, I think I may have made a mistake when transcribing my result from Mathematica into R. My result should be the same as eq. (22) in our notes. But I think this isn't true for this R code. I think the code should read:

```r
probability <- (1 / (foi[i] + mu)) * (foi[i] + exp(-(foi[i] + mu)) * (probability * (foi[i] + mu) - foi[i]))
```

Can you check that you agree (I've simplified the expression here and probably should do the same in the notes)?

File: R/seroprevalence_data.R [hunk 2]
```diff
@@ -139,6 +139,32 @@ prepare_bin_data <- function(serodata) {
   return(xx)
 }
 
+#' Function that returns the probability of being seropositive for age-varying
+#' FoI including seroreversion
```
Comment (ben18785) [inline R/seroprevalence_data.R:143]: force-of-infection (I think good to not use acronyms so people don't get confused)

File: R/seroprevalence_data.R [hunk 3]
```diff
@@ -139,6 +139,32 @@ prepare_bin_data <- function(serodata) {
...
+    probability <-
+      (1 / (foi[i] + mu)) * exp(- (foi[i] + mu)) *
+      (foi[i] * (exp(foi[i] + mu)  - 1) + probability * (foi[i] + mu)
+      )
+  }
+  return(probability)
+}
+
 #' Function that generates the probabilities of being previously exposed to a
 #' pathogen given a historical Force-of-Infection.
```
Comment (ben18785) [inline R/seroprevalence_data.R:169]: lowercase "force-of-infection" since it's not a name

File: R/seroprevalence_data.R [hunk 4]
```diff
@@ -148,7 +174,12 @@ prepare_bin_data <- function(serodata) {
 #'   \item{`tsur`}{Year of the survey}
 #' }
 #' @param foi Numeric atomic vector corresponding to the desired
-#'   Force-of-Infection ordered from past to present
+#' time-varying or age-varying Force-of-Infection to simulate from
```
Comment (ben18785) [inline R/seroprevalence_data.R:177]: Same

File: R/seroprevalence_data.R [hunk 5]
```diff
@@ -148,7 +174,12 @@ prepare_bin_data <- function(serodata) {
...
+#' @param mu Seroreversion rate
+#' @param is_time_varying Boolean specifying whether `foi`
+#' is a time-varying function
+#' @param is_age_varying Boolean specifying whether `foi`
+#' is an age-varying function
```
Comment (ben18785) [inline R/seroprevalence_data.R:182]: I wonder: would these be better off as a single parameter, `model_type = c('age', 'time', 'age-time')`? I think that would resolve the ambiguity you have to handle and easily allow generalisation to other types of FOIs.

File: R/seroprevalence_data.R [hunk 6]
```diff
@@ -164,13 +195,55 @@ prepare_bin_data <- function(serodata) {
...
+    )
+    is_time_varying <- FALSE
+  }
+  if (!any(is_time_varying, is_age_varying)) {
+      warning(
+        "Type of model not specified. ",
+        "Setting `is_time_varying = TRUE`."
+        )
+      is_time_varying <- TRUE
+    }
```
Comment (ben18785) [inline R/seroprevalence_data.R:220]: Think these would not be needed with the above suggested change to the function arguments

File: R/seroprevalence_data.R (hunk 2 above)
Comment (ben18785) [reply R/seroprevalence_data.R:143]: Also I think no need to start function docs with "Function that..." since almost everything we document is a function. Instead, I would reword this as "Calculate probability that an individual of known age is seropositive given an age-varying force-of-infection"

File: R/seroprevalence_data.R (hunk 3 above)
Comment (ben18785) [reply R/seroprevalence_data.R:169]: Also, no need to start off function with "Function" here

File: R/seroprevalence_data.R [hunk 7]
```diff
@@ -211,8 +284,17 @@ get_sim_probability <- function(sim_data, foi) {
 get_sim_n_seropositive <- function(sim_data,
                                    foi,
```
Comment (ben18785) [inline R/seroprevalence_data.R:285]: No need to use "Function..." to describe this function. I would instead say, "Simulate conducting a serosurvey"

File: R/seroprevalence_data.R [hunk 8]
```diff
@@ -211,8 +284,17 @@ get_sim_probability <- function(sim_data, foi) {
...
                                    sample_size_by_age,
+                                   mu = 0,
+                                   is_time_varying = TRUE,
+                                   is_age_varying = FALSE,
```
Comment (ben18785) [inline R/seroprevalence_data.R:289]: See note above about function args

**Review by ben18785 — CHANGES_REQUESTED** [review]
Thanks @ntorresd -- really nice. Apologies, but I think I may have made a mistake with my R code for simulating with seroreversion. (Probably I need to check Stan code too.) I've suggested a change to the R code but please can you check you agree with it?

Something which obs we still need are tests. In terms of analytical results to compare against, I think the lambda = constant case is worth comparing, when the proportion seropositive is given by:

`1/(lambda + mu) (1 - exp(-(lambda + mu)t))`

I'm not sure on reflection whether it's worth checking against the (say) linearly increasing FOI since we know they'll differ... If you wanted to go down that route, you could assume lambda(t) = lambda_0 t, then it is possible to analytically solve the system and results in code like the below for the solution (I used ChatGPT to convert my Mathematica output, so could be wrong); alternatively, you could numerically solve the system using DSolve:

```r
probability <- ((exp(-((t^2 * λ0)/2) - t * μ - μ^2/(2 * λ0)) * (-2 * exp(μ^2/(2 * λ0)) * sqrt(λ0) + 
            2 * exp((t^2 * λ0)/2 + t * μ + μ^2/(2 * λ0)) * sqrt(λ0) + 
            2 * exp(μ^2/(2 * λ0)) * x0 * sqrt(λ0) + 
            sqrt(2 * pi) * μ * erf(μ/(sqrt(2) * sqrt(λ0))) - 
            sqrt(2 * pi) * μ * erf((t * λ0 + μ)/(sqrt(2) * sqrt(λ0)))))/(2 * sqrt(λ0)))

```

File: R/seroprevalence_data.R (hunk 1 above)
Comment (ekamau) [reply R/seroprevalence_data.R:163]: @ben18785 - somehow this new code works the same as what we had previously .. and computes probability the same as the solution we have in the Stan code.

age_foi_sim_soln

**Comment by ekamau** [comment]
On this: https://github.com/epiverse-trace/serofoi/blob/bc1902e2cda3fcc69666627eebecd2c70d22c8d9/R/seroprevalence_data.R#L224

@ntorresd - I could simulate data with seroreversion rate being > 0, i.e., assuming there's seroreversion. Have you been unable to do this?

The solution you have in the code works both when mu = 0 and when mu != 0

**Comment by ekamau** [comment]
@ntorresd, @ben18785 - for this probability function: https://github.com/epiverse-trace/serofoi/blob/bc1902e2cda3fcc69666627eebecd2c70d22c8d9/R/seroprevalence_data.R#L151

you can check or test it against the solution that assumes constant FOI with seroreversion:
```
constant_FOI_solution <- function(age, foi, mu) {
  #foi * (1 - exp(-a * (foi + mu))) / (foi + mu) ## this line works the same as below ..
  (foi/(foi + mu)) * (1 - exp(-a * (foi + mu))) 
}
```   
I have checked, and the constant FOI function works with and w/out seroreversion, i.e., mu == 0 and mu != 0,
and gives the same output/simulations as the age FOI function when foi is a single fixed value ...

let me know if this makes sense - assuming that what we were discussing ...

**Comment by ekamau** [comment]
Adding the code I was using to test the function(s) - might be helpful:
```
library(tidyverse)
library(rstan)
rstan_options(auto_write = TRUE)
options(mc.cores = 4)

simulate_foi_age_exact <- function(age, fois, chunks, mu) {
  I <- 0
  # solves ODE exactly within pieces
  for(i in 1:age) {
    lambda <- fois[chunks[i]]
    I <- (1 / (lambda + mu)) * (lambda + exp(-(lambda + mu)) * (I * (lambda + mu) - lambda))
  }
  I
}

simulate_foi_age_exact2 <- function(age, fois, chunks, mu) {
  I <- 0
  # solves ODE exactly within pieces
  for(i in 1:age) {
    lambda <- fois[chunks[i]]
    I <- (1 / (lambda + mu)) * exp(-(lambda + mu)) * (lambda * (exp(lambda + mu) - 1) + I * (lambda + mu))
  }
  I
}

ages <- seq(1, 40, 1)

#-----------------------------------------------------------------------------------------------
# (1) Model without seroreversion
# generate synthetic data assuming we have data for all individuals aged between 1-n
mu <- 0.0 # no seroreversion
fois <- c(0.01, 0.05, 0.1, 0.03)
chunks <- unlist(map(seq(1, 4, 1), ~rep(., 10))) # how many FOIs to estimate
prob_infected_simulated <- map_dbl(ages, ~simulate_foi_age_exact(., fois, chunks, mu))
prob_infected_simulated2 <- map_dbl(ages, ~simulate_foi_age_exact2(., fois, chunks, mu))

tibble(age=ages, R_function=prob_infected_simulated, R_function2=prob_infected_simulated2) %>%
  pivot_longer(-age) %>%
  ggplot(aes(x=age, y=value, colour=name)) +
  geom_line() + geom_point() + labs(y = "Probability infected") + guides(colour = "none") + 
  facet_wrap(.~ name)


# checking with constant FOI
constant_solution <- function(a, foi, mu) {
  #foi * (1 - exp(-a * (foi + mu))) / (foi + mu) # Ben's solution ..
  (foi/(foi + mu)) * (1 - exp(-a * (foi + mu))) # what's usually published
}

prob_infected_cst <- map_dbl(ages, ~constant_solution(., fois[1], mu))

fois <- rep(fois[1], length(ages)) # constant - for each age!

prob_infected_simulated <- map_dbl(ages, ~simulate_foi_age_exact(., fois, chunks, mu))
prob_infected_simulated2 <- map_dbl(ages, ~simulate_foi_age_exact2(., fois, chunks, mu))

tibble(age=ages, R_function=prob_infected_simulated, R_function2=prob_infected_simulated2,
       cst_function=prob_infected_cst) %>%
  pivot_longer(-age) %>%
  ggplot(aes(x=age, y=value, colour=name)) +
  geom_line() + geom_point() + labs(y = "Probability infected") + guides(colour = "none") + 
  facet_wrap(.~ name)


#-----------------------------------------------------------------------------------------------
#2) With seroreversion:
mu <- 0.1 # no seroreversion
fois <- c(0.01, 0.05, 0.1, 0.03)
chunks <- unlist(map(seq(1, 4, 1), ~rep(., 10))) # how many FOIs to estimate

prob_infected_simulated <- map_dbl(ages, ~simulate_foi_age_exact(., fois, chunks, mu))
prob_infected_simulated2 <- map_dbl(ages, ~simulate_foi_age_exact2(., fois, chunks, mu))

tibble(age=ages, R_function=prob_infected_simulated, R_function2=prob_infected_simulated2) %>%
  pivot_longer(-age) %>%
  ggplot(aes(x=age, y=value, colour=name)) +
  geom_line() + geom_point() + labs(y = "Probability infected") + guides(colour = "none") + 
  facet_wrap(.~ name)


# checking with constant FOI
constant_solution <- function(a, foi, mu) {
  #foi * (1 - exp(-a * (foi + mu))) / (foi + mu) # Ben's solution ..
  (foi/(foi + mu)) * (1 - exp(-a * (foi + mu))) # what's usually published
}

prob_infected_cst <- map_dbl(ages, ~constant_solution(., fois[1], mu))

fois <- rep(fois[1], length(ages)) # constant - for each age!

prob_infected_simulated <- map_dbl(ages, ~simulate_foi_age_exact(., fois, chunks, mu))
prob_infected_simulated2 <- map_dbl(ages, ~simulate_foi_age_exact2(., fois, chunks, mu))

tibble(age=ages, R_function=prob_infected_simulated, R_function2=prob_infected_simulated2,
       cst_function=prob_infected_cst) %>%
  pivot_longer(-age) %>%
  ggplot(aes(x=age, y=value, colour=name)) +
  geom_line() + geom_point() + labs(y = "Probability infected") + guides(colour = "none") + 
  facet_wrap(.~ name)
```

File: R/seroprevalence_data.R (hunk 1 above)
Comment (ntorresd) [reply R/seroprevalence_data.R:163]: It seems like they're all the same. I rewrote it as:

```math
X(t) = X_0 e^{-(\lambda + \mu)} + \frac{\lambda}{\lambda + \mu} (1 - e^{-(\lambda + \mu)}),
```
which is equivalent to what you originally coded:

```math
X(t) = \frac{e^{-(\lambda + \mu)}}{\lambda + \mu} \left( \lambda (e^{\lambda + \mu} - 1) + X_0 (\lambda + \mu) \right).
```
You must have factorized the exponential and the denominator to simplify. Introducing the exponential in the parenthesis and factorizing it inside (leaving out $\lambda$) yields your current suggestion:

```math
X(t) = \frac{1}{\lambda + \mu} \left(  \lambda + e^{-(\lambda + \mu)} (X_0 (\lambda + \mu) - \lambda) \right)
```

I computed them with $\mu = 0.1$ and they all coincide anyway:
image

I think the last one is better just because it uses only one exponential,  but please let me know which version seems less convoluted to you.

Btw, I checked that your solution actually solves the ODE by direct derivation just in case. I haven't gone through the whole demonstration assuming yearly constant FOI just yet, but I understand better your argument now.

File: R/seroprevalence_data.R (hunk 2 above)
Comment (ntorresd) [reply R/seroprevalence_data.R:143]: I'll address this in a separate issue (#163) since it affects other functions other than the data simulation functions.

File: R/seroprevalence_data.R (hunk 3 above)
Comment (ntorresd) [reply R/seroprevalence_data.R:169]: Same as above (see #164).

File: R/seroprevalence_data.R (hunk 4 above)
Comment (ntorresd) [reply R/seroprevalence_data.R:177]: Same as above (see #164).

File: R/seroprevalence_data.R [hunk 9]
```diff
@@ -139,6 +139,32 @@ prepare_bin_data <- function(serodata) {
...
+#' Function that returns the probability of being seropositive for age-varying
+#' FoI including seroreversion
+#'
+#' @param exposure_age Integer corresponding to the age of the exposed cohort
+#' @param foi Numeric atomic vector corresponding to the age-varying FoI to
+#' simulate from
+#' @param mu Seroreversion rate
+#' @return probability of being seropositive for age-varying FoI
+#' including seroreversion
+get_probability_exact_av <- function(
```
[the remark this answers is not in the document]
Comment (ntorresd) [reply R/seroprevalence_data.R:151]: I'll leave it as `probability_exact_age_varying` for lintr not to complain about the length of the function name

**Comment by ntorresd** [comment]
> @ntorresd, @ben18785 - for this probability function:
> 
> https://github.com/epiverse-trace/serofoi/blob/bc1902e2cda3fcc69666627eebecd2c70d22c8d9/R/seroprevalence_data.R#L151
> 
> you can check or test it against the solution that assumes constant FOI with seroreversion:
> 
> ```
> constant_FOI_solution <- function(age, foi, mu) {
>   #foi * (1 - exp(-a * (foi + mu))) / (foi + mu) ## this line works the same as below ..
>   (foi/(foi + mu)) * (1 - exp(-a * (foi + mu))) 
> }
> ```
> 
> I have checked, and the constant FOI function works with and w/out seroreversion, i.e., mu == 0 and mu != 0, and gives the same output/simulations as the age FOI function when foi is a single fixed value ...
> 
> let me know if this makes sense - assuming that's what we were discussing ...

You're right, the functions work pretty well! 

I added a valued based test based on your R code for large sample sizes. I also corrected the way we were testing all the data simulation functions, so I recommend you take a look at https://github.com/epiverse-trace/serofoi/pull/158/commits/5c72282c26de2397c5359dc43ad790decfd738e8 @ben18785 and @ekamau. 

Right now I'm choosing the tolerance to compare the simulated seroprevalence values with the exact probabilities in a rather arbitrary way... I tried doing something similar to what I did for the modelling module (see https://github.com/epiverse-trace/serofoi/pull/153/commits/22244db2e6c5b7873112f33aab6bd3a2997f1c76), where I use as a tolerance half the size of the credible interval; I tried to use here the binomial confidence interval as a criteria, but weirdly enough the tests always passed for arbitrary sample size when I do it (which is nice, but is not very useful for testing), so I chose a value (`tolerance = 10^-3`) that gives good results for a large sample size (10^7) and fails for small sample size (~100).

File: R/seroprevalence_data.R (hunk 1 above)
Comment (ben18785) [reply R/seroprevalence_data.R:162]: Great, thanks @ekamau and @ntorresd and apologies for my convoluted thinking here. I agree that the one with only the single exponential is best.

File: R/seroprevalence_data.R (hunk 2 above)
Comment (ben18785) [reply R/seroprevalence_data.R]: Makes sense! Thanks

File: R/seroprevalence_data.R (hunk 3 above)
Comment (ben18785) [reply R/seroprevalence_data.R:168]: Cool, makes sense

File: R/seroprevalence_data.R (hunk 4 above)
Comment (ben18785) [reply R/seroprevalence_data.R:176]: Cool

File: R/seroprevalence_data.R [hunk 10]
```diff
@@ -139,6 +139,31 @@ prepare_bin_data <- function(serodata) {
...
+    foi,
+    mu = 0
+) {
+  probability <- 0
+  # solves ODE exactly within pieces
+  for (i in 1:age) {
+    probability <-
+      (1 / (foi[i] + mu)) * exp(- (foi[i] + mu)) *
+      (foi[i] * (exp(foi[i] + mu)  - 1) + probability * (foi[i] + mu)
+      )
```
Comment (ben18785) [inline R/seroprevalence_data.R:162]: I'm a bit confused -- have you changed this to the new version (with only a single exponential)?

File: R/seroprevalence_data.R [hunk 11]
```diff
@@ -164,16 +192,39 @@ prepare_bin_data <- function(serodata) {
...
   cohort_ages <- get_cohort_ages(sim_data)
-  exposure_ages <- rev(cohort_ages$age)
-  exposure_matrix <- get_exposure_matrix(sim_data) # nolint: object_usage_linter
-  probabilities <- 1 - exp(-drop(exposure_matrix %*% foi))
+  ages <- rev(cohort_ages$age)
+
+  if (model_type == "time-varying") {
+    exposure_matrix <- get_exposure_matrix(sim_data) # nolint: object_usage_linter
+    probabilities <-
+    (foi / (foi + mu)) * (1 - exp(-drop(exposure_matrix %*% (foi + mu))))
```
Comment (ben18785) [inline R/seroprevalence_data.R:217]: I would put this into its own function (like we've done for age_varying) which gets its own unit tests; it also makes the code more humanly readable.

**Review by ben18785 — DISMISSED** [review]
Thanks @ntorresd -- looks good. A few more comments but quite minor

File: R/seroprevalence_data.R (hunk 11 above)
Comment (ntorresd) [reply R/seroprevalence_data.R:217]: Good point. I added `probability_exact_time_varying` to compute this probability. I'm computing the exposure matrix internally in the function to avoid the use of `get_exposure_matrix` and simplify the function usage.

**Comment by ekamau** [comment]
@ntorresd - thanks for making the changes. it looks good; no further comments at this stage.

File: R/seroprevalence_data.R
Comment (ekamau) [inline R/seroprevalence_data.R:1]: @ntorresd - Have you tested the functions `get_age_group` and `group_sim_data` to see if they are working as expected?

**Review by ekamau — COMMENTED** [review]
looks good ... see my comment / question on age grouping.

File: R/seroprevalence_data.R
Comment (ntorresd) [reply R/seroprevalence_data.R:1]: @ekamau - Yes, I have and they seem to be working fine for data that is not already group and without missing ages. Have you found an example for which they don't work correctly?

**Review by ekamau — APPROVED** [review]
all good!
